secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,244 @@
+struct RecoveryPhraseStepFulfillment: Equatable {
+    var groupIndex: Int
+    var word: String
+}
+
+struct RecoveryPhraseValidationState: Equatable {
+    enum Step: Equatable {
+        case initial
+        case incomplete
+        case complete
+    }
+
+    enum Route: Equatable, CaseIterable {
+        case success
+        case failure
+    }
+
+    static let wordGroupSize = 6
+    static let phraseChunks = 4
+
+    var phrase: RecoveryPhrase
+    var missingIndices: [Int]
+    var missingWordChips: [PhraseChip.Kind]
+    var fulfillments: [RecoveryPhraseStepFulfillment]
+    var route: Route?
+    var step: Step {
+        guard !fulfillments.isEmpty else {
+            return  .initial
+        }
+
+        guard fulfillments.count >= missingIndices.count else {
+            return .incomplete
+        }
+
+        return .complete
+    }
+
+    var isValid: Bool {
+        self.resultingPhrase == phrase.words
+    }
+}
+
+extension RecoveryPhraseValidationViewStore {
+    func bindingForRoute(_ route: RecoveryPhraseValidationState.Route) -> Binding<Bool> {
+        self.binding(
+            get: { $0.route == route },
+            send: { isActive in
+                return .updateRoute(isActive ? route : nil)
+            }
+        )
+    }
+}
+
+
+extension RecoveryPhraseValidationState {
+    /// creates an initial `RecoveryPhraseValidationState` with no completions and random missing indices.
+    /// - Note: Use this function to create a random validation puzzle for a given phrase.
+    static func random(phrase: RecoveryPhrase) -> Self {
+        let missingIndices = Self.randomIndices()
+        let missingWordChipKind = Self.pickWordsFromMissingIndices(indices: missingIndices, phrase: phrase).shuffled()
+        return RecoveryPhraseValidationState(
+            phrase: phrase,
+            missingIndices: missingIndices,
+            missingWordChips: missingWordChipKind,
+            fulfillments: []
+        )
+    }
+}
+
+extension RecoveryPhraseValidationState {
+    /// Given an array of RecoveryPhraseStepCompletion, missing indices, original phrase and the number of groups it was split into,
+    /// assembly the resulting phrase. This comes up with the "proposed solution" for the recovery phrase validation challenge.
+    /// - returns:an array of String containing the recovery phrase words ordered by the original phrase order.
+    var resultingPhrase: [String] {
+        precondition(missingIndices.count == fulfillments.count)
+        precondition(fulfillments.count == Self.phraseChunks)
+
+        var words = phrase.words
+        let groupLength = words.count / Self.phraseChunks
+        // iterate based on the completions the user did on the UI
+        for wordCompletion in fulfillments {
+            // figure out which phrase group (chunk) this completion belongs to
+            let i = wordCompletion.groupIndex
```

Comment:
No, I have no swift lint warnings. Maybe it's not being called? That's weird!

Reply:
Looks like it was added to the exclusion list, we should remove that in a different PR.

Reply:
what? i? the most widely known abbreviation for indices?   The use of `i` as index is no code smell.

Reply:
It's not code smell these are just the standard Swift guidelines that most every project would use.  Single character variables are widely discouraged in all cases.

Reply:
I noticed in reviewing the rest of this PR that we do use the term index and indices elsewhere.  I still think it's a good idea to do that here as well.  It fits better with Swifts naming conventions and style guides, and it keeps the whole project consistent.

Reply:
I changed it to groupIndex since it doesn't hurt. 

I don't believe that single character variable names are wrong. They have their proper place and usage. Using i and j or x and y for matrices is an example of proper use.

Reply:
x, y and z are common for describing coordinates or measurements along axis' that's very true, but other usages are discouraged.  This is probably underscored by the fact that the default swiftlint configuration disallows them altogether.